crates/node/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+pub mod cluster_controller;
+mod options;
+pub mod worker;
+
+use codederror::CodedError;
+use futures::future::OptionFuture;
+use restate_types::identifiers::NodeId;
+use std::convert::Infallible;
+use std::str::FromStr;
+use std::time::Duration;
+use tokio::task::JoinError;
+use tonic::codegen::http::uri::InvalidUri;
+use tonic::transport::{Channel, Uri};
+use tracing::{info, instrument};
+
+use crate::cluster_controller::ClusterControllerRole;
+use crate::worker::WorkerRole;
+pub use options::{Options, OptionsBuilder as NodeOptionsBuilder};
+pub use restate_admin::OptionsBuilder as AdminOptionsBuilder;
+use restate_cluster_controller::proto::cluster_controller_client::ClusterControllerClient;
+use restate_cluster_controller::proto::AttachmentRequest;
+pub use restate_meta::OptionsBuilder as MetaOptionsBuilder;
+use restate_types::retries::RetryPolicy;
+pub use restate_worker::{OptionsBuilder as WorkerOptionsBuilder, RocksdbOptionsBuilder};
+
+#[derive(Debug, thiserror::Error, CodedError)]
+pub enum Error {
+    #[error("worker failed: {0}")]
+    Worker(
+        #[from]
+        #[code]
+        worker::WorkerRoleError,
+    ),
+    #[error("controller failed: {0}")]
+    Controller(
+        #[from]
+        #[code]
+        cluster_controller::ClusterControllerRoleError,
+    ),
+    #[error("failed to attach to cluster at '{0}': {1}")]
+    #[code(unknown)]
+    Attachment(Uri, tonic::Status),
+    #[error("component '{0}' panicked: {1}")]
+    #[code(unknown)]
+    Panic(&'static str, JoinError),
+}
+
+impl Error {
+    fn panic(component: &'static str, err: JoinError) -> Self {
+        Error::Panic(component, err)
+    }
+}
+
+#[derive(Debug, thiserror::Error, CodedError)]
+pub enum BuildError {
+    #[error("building worker failed: {0}")]
+    Worker(
+        #[from]
+        #[code]
+        worker::WorkerRoleBuildError,
+    ),
+    #[error("invalid controller endpoint: {0}")]
+    #[code(unknown)]
+    InvalidControllerEndpoint(#[from] InvalidUri),
+}
+
+pub struct Node {
+    node_id: NodeId,
+    cluster_controller_endpoint: Uri,
+
+    cluster_controller_role: Option<ClusterControllerRole>,
+    worker_role: WorkerRole,
+}
+
+impl Node {
+    pub fn new(
+        node_id: impl Into<NodeId>,
+        cluster_controller_location: ClusterControllerLocation,
+        options: Options,
+    ) -> Result<Self, BuildError> {
+        let (cluster_controller_role, cluster_controller_endpoint) =
+            match cluster_controller_location {
+                ClusterControllerLocation::Local => {
```

Comment:
I wonder if we can remove this specialization. A local cluster controller might as simple as defaulting the address to localhost:5123 (or a unix domain socket path).

This can be in a follow up PR of course.

Reply:
Yes, I think you are right. Right now I kind of encode the role via this enum which will be replaced via the `NodeConfig` in https://github.com/restatedev/restate/pull/1168. I will clean it up once we have the `NodeConfig` merged.